### DISCUSSION TITLE
Systembridge going offline

### DISCUSSION THREAD

**Discussion by aniceberg** [timmo001/system-bridge#2858] [Help]
I see that Systembridge keeps going offline after a while on my RPi.

I grabbed the systemctl status and got the below python code errors if this is helpful for diagnosing the issue. Are there other logs I should be pulling up to further diagnose the issue and which ones?

```
● systembridge.service - Systembridge
     Loaded: loaded (/etc/systemd/system/systembridge.service; enabled; vendor preset: enabled)
     Active: active (exited) since Thu 2023-11-16 18:40:42 EST; 16h ago
    Process: 555 ExecStart=/usr/bin/python3.9 -m systembridgebackend --no-gui (code=exited, status=0/SUCCESS)
   Main PID: 555 (code=exited, status=0/SUCCESS)
        CPU: 5.907s

Nov 16 18:40:50 nvr python3.9[555]:   File "/usr/local/lib/python3.9/dist-packages/systembridgebackend/__init__.py", line 57, in __init__
Nov 16 18:40:50 nvr python3.9[555]:     self._server = Server(
Nov 16 18:40:50 nvr python3.9[555]:   File "/usr/local/lib/python3.9/dist-packages/systembridgebackend/server/__init__.py", line 72, in __init__
Nov 16 18:40:50 nvr python3.9[555]:     self._mdns_advertisement.advertise_server()
Nov 16 18:40:50 nvr python3.9[555]:   File "/usr/local/lib/python3.9/dist-packages/systembridgebackend/server/mdns.py", line 42, in advertise_server
Nov 16 18:40:50 nvr python3.9[555]:     info = ServiceInfo(
Nov 16 18:40:50 nvr python3.9[555]:   File "src/zeroconf/_services/info.py", line 189, in zeroconf._services.info.ServiceInfo.__init__
Nov 16 18:40:50 nvr python3.9[555]:   File "/usr/local/lib/python3.9/dist-packages/zeroconf/_utils/net.py", line 70, in _encode_address
Nov 16 18:40:50 nvr python3.9[555]:     return socket.inet_pton(address_family, address)
Nov 16 18:40:50 nvr python3.9[555]: OSError: illegal IP address string passed to inet_pton
```

**Comment by aniceberg** [comment]
Same thing happening with 3.11.2, it starts and shuts after a few seconds


```
pi@nvr:~ $ sudo systemctl status systembridge
● systembridge.service - Systembridge
     Loaded: loaded (/etc/systemd/system/systembridge.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2023-11-18 20:46:38 EST; 1s ago
   Main PID: 6447 (python3.9)
      Tasks: 1 (limit: 3723)
        CPU: 1.011s
     CGroup: /system.slice/systembridge.service
             └─6447 /usr/bin/python3.9 -m systembridgebackend --no-gui

Nov 18 20:46:39 nvr python3.9[6447]:   File "/usr/local/lib/python3.9/dist-packages/pydantic/main.py", line 178, in __new__
Nov 18 20:46:39 nvr python3.9[6447]:     annotations = resolve_annotations(namespace.get('__annotations__', {}), namespace.get('__module__', None))
Nov 18 20:46:39 nvr python3.9[6447]:   File "/usr/local/lib/python3.9/dist-packages/pydantic/typing.py", line 400, in resolve_annotations
Nov 18 20:46:39 nvr python3.9[6447]:     value = _eval_type(value, base_globals, None)
Nov 18 20:46:39 nvr python3.9[6447]:   File "/usr/lib/python3.9/typing.py", line 283, in _eval_type
Nov 18 20:46:39 nvr python3.9[6447]:     return t._evaluate(globalns, localns, recursive_guard)
Nov 18 20:46:39 nvr python3.9[6447]:   File "/usr/lib/python3.9/typing.py", line 539, in _evaluate
Nov 18 20:46:39 nvr python3.9[6447]:     eval(self.__forward_code__, globalns, localns),
Nov 18 20:46:39 nvr python3.9[6447]:   File "<string>", line 1, in <module>
Nov 18 20:46:39 nvr python3.9[6447]: TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
pi@nvr:~ $ sudo systemctl status systembridge
● systembridge.service - Systembridge
     Loaded: loaded (/etc/systemd/system/systembridge.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Sat 2023-11-18 20:46:44 EST; 4s ago
    Process: 6485 ExecStart=/usr/bin/python3.9 -m systembridgebackend --no-gui (code=exited, status=1/FAILURE)
   Main PID: 6485 (code=exited, status=1/FAILURE)
        CPU: 1.146s

Nov 18 20:46:44 nvr systemd[1]: systembridge.service: Scheduled restart job, restart counter is at 5.
Nov 18 20:46:44 nvr systemd[1]: Stopped Systembridge.
Nov 18 20:46:44 nvr systemd[1]: systembridge.service: Consumed 1.146s CPU time.
Nov 18 20:46:44 nvr systemd[1]: systembridge.service: Start request repeated too quickly.
Nov 18 20:46:44 nvr systemd[1]: systembridge.service: Failed with result 'exit-code'.
Nov 18 20:46:44 nvr systemd[1]: Failed to start Systembridge.
pi@nvr:~ $
```

**Reply by timmo001** [reply]
Please update to the latest version of python. 3.9 is no longer supported

**Reply by timmo001** [reply]
Using the binary is the simplest way to use system bridge. See the installation docs here:

https://system-bridge.timmo.dev/docs/install

**Reply by aniceberg** [reply]
Thanks. Any idea which binary release would run on a RPi 4 running Raspian Buster, a distro based on debian?

**Reply by aniceberg** [reply]
Here is my issue-- AFAIK, there's no binary that runs on Raspbian Buster/Bullseye which I can run using the elevated privileges that Systembridge requires on the host machine.

When I go the Python route, I can install it (`python -m pip install --upgrade systembridgeshared systembridgebackend systembridgecli`) and run it (`/usr/bin/python -m systembridgebackend --no-gui`) without issue however if I send a keypress I get this in the log:
```
2023-11-27 10:07:02 INFO (MainThread) [systembridgebackend.server.websocket.WebSocketHandler] Received: KEYBOARD_KEYPRESS
2023-11-27 10:07:02 ERROR (MainThread) [systembridgebackend.server.websocket.WebSocketHandler] You must be root to use this library on linux.
```

However I don't know how to run python as the root user- because when I attempt that, I get this message:
```
sudo /usr/bin/python -m systembridgebackend --no-gui
/usr/bin/python: No module named systembridgebackend
[5]+  Killed                  /usr/bin/python -m systembridgebackend --no-gui
```

I tried installing the python modules under the root user (`sudo python -m pip install --upgrade systembridgeshared systembridgebackend systembridgecli`) but it gets stuck when trying to install the cryptography module (because of rustc dependencies, which in turn is having issues because it can't find the rustc compiler under the root user.)

>       error: rustup could not choose a version of rustc to run, because one wasn't specified explicitly, and no default is configured.
>       help: run 'rustup default stable' to download the latest stable release of Rust and set it as your default toolchain.
>       
>           =============================DEBUG ASSISTANCE=============================
>           If you are seeing a compilation error please try the following steps to
>           successfully install cryptography:
>           1) Upgrade to the latest pip and try again. This will fix errors for most
>              users. See: https://pip.pypa.io/en/stable/installing/#upgrading-pip
>           2) Read https://cryptography.io/en/latest/installation/ for specific
>              instructions for your platform.
>           3) Check our frequently asked questions for more information:
>              https://cryptography.io/en/latest/faq/
>           4) Ensure you have a recent Rust toolchain installed:
>              https://cryptography.io/en/latest/installation/#rust
>       
>           Python: 3.12.0
>           platform: Linux-6.1.21-v8+-aarch64-with-glibc2.31
>           pip: n/a
>           setuptools: 69.0.2
>           setuptools_rust: 1.8.1
>           rustc: n/a
>           =============================DEBUG ASSISTANCE=============================
>       
>       error: can't find Rust compiler



As you can tell, I'm not that python or linux savvy in general-- do you have any suggestions or solutions?
